### ISSUE TITLE
Matrix multiplication on GPU in JAX is 4 times slower than in PyTorch 

### ISSUE THREAD

**Issue by dfdx** [google/jax#18662]
### Description

I originally posted it as a [discussion](https://github.com/google/jax/discussions/18596), but now I have a small reproducible example, so re-posting it as a bug.

I test multiplication of 2 matrices on GPU:

```python
import jax
rng = jax.random.PRNGKey(0)
x = jax.random.normal(rng, (8, 4096))
w = jax.random.normal(rng, (4096, 1024))
```
Also jitted version:

```python
def f(x, w):
    return x @ w

jit_f = jax.jit(f)
```
And compare it to the PyTorch version:

```python
import torch
import numpy as np
def jax2pt(x):
    return torch.from_numpy(np.asarray(x).copy())

pt_x = jax2pt(x).to(torch.device("cuda"))
pt_w = jax2pt(w).to(torch.device("cuda"))
```
As expected, normal and jitted versions in JAX perform on the same level. However, PyTorch version is somehow 4 times faster:

```python
In [18]: %timeit (x @ w).block_until_ready()
99.3 µs ± 2.98 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [19]: %timeit jit_f(x, w).block_until_ready()
92.9 µs ± 3.32 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [20]: %timeit pt_x @ pt_w
24.7 µs ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
I checked that both JAX arrays are on GPU, but as I mentioned in the [discussion](https://github.com/google/jax/discussions/18596), the code might be actually executed on CPU.

### What jax/jaxlib version are you using?

jax v0.4.20, jaxlib v0.4.20+cuda11.cudnn86, torch v2.1.1

### Which accelerator(s) are you using?

GPU

### Additional system info?

1.26.2 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] uname_result(system='Linux', node='6ccb2569044d', release='5.15.0-1051-azure', version='#59~20.04.1-Ubuntu SMP Tue Oct 17 16:45:08 UTC 2023', machine='x86_64')

### NVIDIA GPU info

```
NVIDIA A100
Driver Version: 525.147.05
CUDA Version: 12.0
Base Docker image: nvidia/cuda:12.0.1-cudnn8-devel-ubuntu22.04
```

**Comment by dfdx** [comment]
Screenshot of the profiler:
![image](https://github.com/google/jax/assets/428171/0347226b-3637-40e3-866c-895af762420f)

**Comment by jjyyxx** [comment]
For a fair comparison, you should wait PyTorch cuda stream as `block_until_ready` do. So, try to use
```python
%timeit pt_x @ pt_w; torch.cuda.synchronize()
```
Even though, on my side, JAX is slower than PyTorch by around two times. I guess the reason is either a different kernel or different dispatch cost?

**Comment by dfdx** [comment]
Indeed, synchronization in PyTorch reduces the difference to 2-2.5x: 

```python
In [16]: %timeit pt_x @ pt_w; torch.cuda.synchronize()
38.7 µs ± 94.3 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
I don't think it's caused by different dispatch because in the larger model the difference in execution speed is even more striking (see the linked discussion). 
Regarding kernel, I assume you are talking about different CUDA kernels. Is there a way in JAX to use a different kernel?

**Comment by jjyyxx** [comment]
Regarding your linked discussion, my "rule of thumb" is to wrap all your computation inside `jax.jit` (e.g., during training, the forward/backward/gradient update are typically all fused in one jit call). Without jit, JAX is undoubtly slower than PyTorch. But from my experience, when jit is properly used, JAX is typically 2x to 1000x faster than PyTorch (obvious for small but complex models, due to reduced dispatch cost and fusion).

I suspect that the problem revealed in this issue is different from that in your discussion (simply caused by missing jit).

**Comment by dfdx** [comment]
And you are right! Setup with a more complex function:

```python
def f(x, w):
    w = w / w.std()      # .std() for better stability
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    x = x @ w / x.std()
    return x

jit_f = jax.jit(f)

x = jax.random.normal(rng, (128, 4096))
w = jax.random.normal(rng, (4096, 4096))
pt_x = jax2pt(x).to(torch.device("cuda"))
pt_w = jax2pt(w).to(torch.device("cuda"))
```
Test:
```python
In [1]: %timeit f(pt_x, pt_w); torch.cuda.synchronize()
4.68 ms ± 1.96 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [2]: %timeit f(x, w).block_until_ready()
3.46 ms ± 279 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [3]: %timeit jit_f(x, w).block_until_ready()
1.51 ms ± 1.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
Thanks for clarification!
